submit-api/src/Cardano/TxSubmit/Tx.hs

```diff
@@ -1,68 +1,112 @@
+{-# LANGUAGE GADTs #-}
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
 
 module Cardano.TxSubmit.Tx
-  ( TxSubmitVar (..)
-  , newTxSubmitVar
-  , readTxSubmit
-  , writeTxSubmitResponse
+  ( TxSubmitError (..)
+  , renderTxSubmitError
   , submitTx
   ) where
 
-import Cardano.Prelude hiding
-    ( atomically )
+import Cardano.Prelude
 
-import Cardano.Chain.Byron.API
-    ( ApplyMempoolPayloadErr (..) )
-import Cardano.Chain.UTxO
-    ( TxId )
-import Cardano.TxSubmit.Types
-import Control.Monad.Class.MonadSTM.Strict
-    ( StrictTMVar, atomically, newEmptyTMVarM, putTMVar, takeTMVar )
+import Cardano.Api.TxSubmit hiding
+    ( submitTx )
+import Cardano.Api.Typed
+    ( Byron
+    , LocalNodeConnectInfo
+    , NodeConsensusMode (..)
+    , Shelley
+    , Tx (..)
+    , TxBody (..)
+    , TxId
+    , getTxId
+    , localNodeConsensusMode
+    )
+import Cardano.TxSubmit.ErrorRender
+    ( renderApplyMempoolPayloadErr, renderEraMismatch )
 import Ouroboros.Consensus.Byron.Ledger
-    ( ByronBlock (..), GenTx (..) )
-import Ouroboros.Network.Protocol.LocalTxSubmission.Type
-    ( SubmitResult (..) )
+    ( ByronBlock )
+import Ouroboros.Consensus.Cardano.Block
+    ( EraMismatch (..)
+    , HardForkApplyTxErr (ApplyTxErrByron, ApplyTxErrShelley, ApplyTxErrWrongEra)
+    )
+import Ouroboros.Consensus.Ledger.SupportsMempool
+    ( ApplyTxErr )
+import Ouroboros.Consensus.Shelley.Ledger
+    ( ShelleyBlock )
+import Ouroboros.Consensus.Shelley.Protocol.Crypto
+    ( TPraosStandardCrypto )
 
--- The type of 'reject' (determined by ouroboros-network) is currently 'Maybe String'.
--- Hopefully that will be fixed to make it a concrete type.
--- See: https://github.com/input-output-hk/ouroboros-network/issues/1335
-data TxSubmitVar = TxSubmitVar
-  { txSubmit :: !(StrictTMVar IO (GenTx ByronBlock))
-  , txRespond :: !(StrictTMVar IO (SubmitResult ApplyMempoolPayloadErr))
-  }
+import qualified Cardano.Api.TxSubmit as Api
+import qualified Cardano.Chain.UTxO as Byron
+import qualified Shelley.Spec.Ledger.BaseTypes as Shelley
+import qualified Shelley.Spec.Ledger.Tx as Shelley
 
-newTxSubmitVar :: IO (TxSubmitVar)
-newTxSubmitVar =
-  TxSubmitVar <$> newEmptyTMVarM <*> newEmptyTMVarM
+-- | An error that can occur while submitting a transaction to a local node.
+data TxSubmitError
+  = TxSubmitByronError !(ApplyTxErr ByronBlock)
+  | TxSubmitShelleyError !(ApplyTxErr (ShelleyBlock TPraosStandardCrypto))
+  | TxSubmitEraMismatchError !EraMismatch
+  deriving (Eq, Show)
 
--- | Read a previously submitted tx from the TMVar.
-readTxSubmit :: TxSubmitVar -> IO (GenTx ByronBlock)
-readTxSubmit tsv =
-  atomically $ takeTMVar (txSubmit tsv)
+renderTxSubmitError :: TxSubmitError -> Text
+renderTxSubmitError tse =
+  case tse of
```

Comment:
Suggestion: use `LambdaCase` here for extra brevity.

```hs
renderTxSubmitError = \case
    TxSubmitByronError err -> renderApplyMempoolPayloadErr err
    TxSubmitShelleyError err -> show err
    TxSubmitEraMismatchError err -> renderEraMismatch err
```

Reply:
Nothing wrong with `LambdaCase`, but it hasn't been used throughout this project and I don't think I'd introduce it just for these simple cases. If you insist that the brevity is worth it, then I'll make the change.

Reply:
> Nothing wrong with `LambdaCase`, but it hasn't been used throughout this project and I don't think I'd introduce it just for these simple cases. If you insist that the brevity is worth it, however, I'll make the change.

Don't worry, it's certainly not essential! If the project style is to not use `LambaCase` then that's fine, in my opinion.

Using `LambdaCase` does one have advantage, IMO: it demonstrates to the reader that the binding (`bse` in this case) is not used in the definition of the function. Of course, the reader can check this for themselves (by scanning the definition manually), but using `LambdaCase` saves the reader a small amount of effort.